pymeasure/instruments/regatron/regatron.py

```diff
@@ -0,0 +1,602 @@
+"""
+from pymeasure.instruments import Instrument
+from pymeasure.adapters.serial import SerialAdapter
+
+import numpy as np
+
+import struct
+import serial
+
+import time
+
+timeout_rise_voltage = 0.1
+timeout_rise_power = 0.1
+
+class Regatron():
+    """
+    Driver for Regatron TopCon compatible Power Supplies
+    """
+    
+    def __init__(self, port_str):            
+        self.port = serial.Serial(port=port_str,baudrate=38400, bytesize=8, parity='N', stopbits=1, timeout=1.0)
+
+    def open(self):
+        # Configure the COM Port
+        if not self.port.is_open:
+            self.port.open()
+        self._serial = self.getSerialNumber()
+        self._fw = self.getFirmwareVersion()
+    
+    def close(self):
+        self.port.close()
+    
+    def getProperties(self):
+        return {
+            'deviceVendor':     'Regatron',
+            'deviceModel':      'TopCon Device',
+            'deviceSerial':     self._serial,
+            'deviceFirmware':   self._fw
+        }
+    
+    def _sendFrame(self, frame):
+        """
+        Send a frame
+        
+        Calculates the checksum and attaches a header to the outgoing
+        data frame.
+        
+        Checksum is calculated by adding all data bytes, then mod by 0x100
+        First byte is a sync byte, it is always 0xA5
+        
+        :returns: None
+        """
+        fmt = 'BBB'
+        
+        total = 0
+        for bt in bytearray(frame):
+            total += bt
+        checksum = total % 0x100
+        
+        header = struct.pack(fmt, 0xA5, len(frame), checksum)
+        
+        full = header + frame
+        
+        self.port.write(full)
+        self.port.flush()
+        
+    def _recvFrame(self):
+        """
+        Receive a frame
+        
+        Reads the header first to find out how many bytes to receive
+        in the payload section
+        
+        :returns: byte string, payload section of frame
+        """
+        header = self.port.read(3)
+        
+        sync, len, checksum = struct.unpack('BBB', header)
+        # TODO Verify checksum?
+        
+        payload = self.port.read(len)
+        
+        return payload
+        
+    def _cmdReadMemoryWord(self, address):
+        """
+        Read a memory word
+        
+        Talk ID: 0x10
+        Address is Little-endian (Least Significant Byte first)
+        
+        :returns: int16
+        """
+        frame = struct.pack('BBBB', 0x10, address&0xFF, (address>>8)&0xFF, (address>>16)&0xFF)
+        
+        self._sendFrame(frame)
+        
+        ret = self._recvFrame()
+        
+        try:
+            talkid, status, datal, datah = struct.unpack('BBBB', ret)
+            
+            # Process returned data
+            data = (datah<<8) | (datal)
+            if data > 32768:
+                data = data-65536;
+            return data
+        
+        except:
+            return 0
+    
+    def _cmdWriteMemoryWord(self, address, data):
```

Comment:
I think this method should be abstracted to an Adapter subclass, to help separate the instrument from the communication method.

Reply:
Same with _cmdReadMemoryWord, _recvFrame, _sendFrame, right?

Reply:
I like that this does frame-based communication (as opposed to the "usual" message-based). @cjermain do you think it feasible/worthwhile to try to find a way to integrate this type of communication into pymeasure, so that control/measurement/setting factories could be used instead of purely methods as in the present PR?

Reply:
Yea, I think it is worthwhile to try to keep the API as consistent as possible.

Reply:
darn it the last comment was still "pending" in the review queue, was made in March.